File: Shared/Blog.FSharp.Core/Helpers/Utilities.fs
namespace Blog.FSharp.Core.Helpers

open System
open System.IO
open System.Security.Cryptography
open System.Text
open System.Threading
open System.Threading.Tasks

/// <summary>
/// System utilities.
/// </summary>
module Utilities =

    /// <summary>
    /// Computes the SHA1 hash of a file, retrying up to 3 times in case of IO exceptions.
    /// </summary>
    /// <param name="filePath">File path.</param>
    /// <returns>Byte array of the hash.</returns>
    let computeHash (filePath: string) : byte[] =
        let mutable runCount = 1
        let mutable result: byte[] = null
        while runCount < 4 do
            try
                if not (File.Exists(filePath)) then
                    raise (FileNotFoundException())
                
                use fs = File.OpenRead(filePath)
                result <- SHA1.Create().ComputeHash(fs)
                runCount <- 4 // exit loop
            with
            | :? IOException as ex ->
                if runCount = 3 || ex.HResult <> -2147024864 then
                    raise ex
                else
                    Thread.Sleep(TimeSpan.FromSeconds(Math.Pow(2.0, float runCount)))
                    runCount <- runCount + 1
        if isNull result then Array.zeroCreate 20 else result

    /// <summary>
    /// Reads the content of a file asynchronously.
    /// </summary>
    /// <param name="filePath">File path.</param>
    /// <returns>File content as string, or null if exception occurs.</returns>
    let getFileContent (filePath: string) : Task<string> = task {
        try
            if not (File.Exists(filePath)) then
                raise (FileNotFoundException())
            return! File.ReadAllTextAsync(filePath)
        with
        | _ -> return null
    }

    /// <summary>
    /// Deletes a file if it exists.
    /// </summary>
    /// <param name="filePath">File path.</param>
    let deleteFile (filePath: string) =
        try
            if not (File.Exists(filePath)) then
                raise (FileNotFoundException())
            File.Delete(filePath)
        with
        | _ -> () // ignore errors

    /// <summary>
    /// Writes content to a file asynchronously, creating folder if necessary.
    /// </summary>
    /// <param name="filePath">File path.</param>
    /// <param name="content">Content string.</param>
    let setFileContent (filePath: string) (content: string) : Task = task {
        try
            let folder = Path.GetDirectoryName(filePath)
            if not (Directory.Exists(folder)) then Directory.CreateDirectory(folder) |> ignore
            use fs = File.Create(filePath)
            let bytes = UTF8Encoding(true).GetBytes(content)
            do! fs.WriteAsync(bytes, 0, bytes.Length)
        with
        | _ -> () // ignore errors
    }

    /// <summary>
    /// Generates an ETag from key and content bytes.
    /// </summary>
    /// <param name="key">Key string.</param>
    /// <param name="contentBytes">Content bytes.</param>
    /// <returns>ETag string.</returns>
    let getETag (key: string) (contentBytes: byte[]) : string =
        let keyBytes = Encoding.UTF8.GetBytes(key)
        let combined = Array.append keyBytes contentBytes
        use md5 = MD5.Create()
        let hash = md5.ComputeHash(combined)
        BitConverter.ToString(hash).Replace("-", String.Empty)

    /// <summary>
    /// Gets the folder of the executing assembly.
    /// </summary>
    /// <returns>Folder path.</returns>
    /// <summary>
    let getWorkingFolder () =
        match System.Reflection.Assembly.GetEntryAssembly() with
        | null -> null
        | asm -> Path.GetDirectoryName(asm.Location)

File: Shared/Blog.Core/Helpers/Utilities.cs
// <copyright file="Utilities.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Blog.Core.Helpers;

using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

/// <summary>
/// System utilities.
/// </summary>
public static class Utilities
{
    /// <summary>
    /// Computes the hash.
    /// </summary>
    /// <param name="filePath">The file path.</param>
    /// <returns>byte[].</returns>
    /// <exception cref="FileNotFoundException">File not found exception.</exception>
    public static byte[] ComputeHash(string filePath)
    {
        var runCount = 1;
        while (runCount < 4)
        {
            try
            {
                if (!File.Exists(filePath))
                {
                    throw new FileNotFoundException();
                }

                using var fs = File.OpenRead(filePath);

                return SHA1.Create().ComputeHash(fs);
            }
            catch (IOException ex)
            {
                if (runCount == 3 || ex.HResult != -2147024864)
                {
                    throw;
                }
                else
                {
                    Thread.Sleep(TimeSpan.FromSeconds(Math.Pow(2, runCount)));
                    runCount++;
                }
            }
        }

        return new byte[20];
    }

    /// <summary>
    /// Gets the content of the file.
    /// </summary>
    /// <param name="filePath">The file path.</param>
    /// <returns>Task.</returns>
    /// <exception cref="FileNotFoundException">File not found exception.</exception>
    public static async Task<string> GetFileContent(string filePath)
    {
        try
        {
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException();
            }

            return await File.ReadAllTextAsync(filePath);
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Remove file from directory (physical deleting).
    /// </summary>
    /// <param name="filePath">The file path.</param>
    /// <exception cref="FileNotFoundException">File not found exception.</exception>
    public static void DeleteFile(string filePath)
    {
        try
        {
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException();
            }

            File.Delete(filePath);
        }
        catch
        {
            // ignored
        }
    }

    /// <summary>
    /// Sets the content of the file.
    /// </summary>
    /// <param name="filePath">The file path.</param>
    /// <param name="content">The content.</param>
    public static async Task SetFileContent(string filePath, string content)
    {
        try
        {
            var folder = Path.GetDirectoryName(filePath);
            if (!string.IsNullOrEmpty(folder) && !Directory.Exists(folder))
            {
                Directory.CreateDirectory(folder);
            }

            if (string.IsNullOrEmpty(filePath) || File.Exists(filePath))
            {
            }

            await using var fs = File.Create(filePath);

            var info = new UTF8Encoding(true).GetBytes(content);
            await fs.WriteAsync(info, 0, info.Length);
        }
        catch
        {
            // ignored
        }
    }

    /// <summary>
    /// Generate ETag for content bytes.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="contentBytes">The content bytes.</param>
    /// <returns>string.</returns>
    public static string GetETag(string key, byte[] contentBytes)
    {
        var keyBytes = Encoding.UTF8.GetBytes(key);
        var combinedBytes = Combine(keyBytes, contentBytes);

        return GenerateETag(combinedBytes);
    }

    /// <summary>
    /// Gets the working folder.
    /// </summary>
    /// <returns>string.</returns>
    public static string GetWorkingFolder()
    {
        var location = System.Reflection.Assembly.GetEntryAssembly()?.Location;

        return Path.GetDirectoryName(location);
    }

    /// <summary>
    /// Generates the e tag.
    /// </summary>
    /// <param name="data">The data.</param>
    /// <returns>string.</returns>
    private static string GenerateETag(byte[] data)
    {
        using var md5 = MD5.Create();

        var hash = md5.ComputeHash(data);
        var hex = BitConverter.ToString(hash);

        return hex.Replace("-", string.Empty);
    }

    /// <summary>
    /// Combines the specified a.
    /// </summary>
    /// <param name="a">a.</param>
    /// <param name="b">The b.</param>
    /// <returns>byte[].</returns>
    private static byte[] Combine(byte[] a, byte[] b)
    {
        var c = new byte[a.Length + b.Length];
        Buffer.BlockCopy(a, 0, c, 0, a.Length);
        Buffer.BlockCopy(b, 0, c, a.Length, b.Length);

        return c;
    }
}